### PR TITLE
py-synthesis-workflow, py-neurocollage: Add synthesis-workflow and all dependencies

### PR DESCRIPTION
**Pull request by adrien-berchet**

### CODE REVIEW

File: bluebrain/repo-bluebrain/packages/py-synthesis-workflow/package.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    depends_on("py-neurocollage@0.3:", type=("build", "run"))
+    depends_on("py-neurom@3.2.2:", type=("build", "run"))
+    depends_on("py-neurots@3.3.1:", type=("build", "run"))
+    depends_on("py-numpy@1.24.1:", type=("build", "run"))
+    depends_on("py-pandas@1.5.3:", type=("build", "run"))
+    depends_on("py-pyyaml@6:", type=("build", "run"))
+    depends_on("py-region-grower@1:", type=("build", "run"))
+    depends_on("py-scipy@1.10:", type=("build", "run"))
+    depends_on("py-seaborn@0.12.2:", type=("build", "run"))
+    depends_on("py-tmd@2.2:", type=("build", "run"))
```
Comment (matz-e) [inline bluebrain/repo-bluebrain/packages/py-synthesis-workflow/package.py]: TMD changed source distribution names recently. Can we require the latest version here?

Comment (adrien-berchet) [reply bluebrain/repo-bluebrain/packages/py-synthesis-workflow/package.py]: Yes sure

**Comment by adrien-berchet** [comment]
Hi @matz-e 
Do you have any idea why the `proj` package fails? I did not change anything to it.

**Comment by matz-e** [comment]
> Hi @matz-e Do you have any idea why the `proj` package fails? I did not change anything to it.

Oops, sorry, fell off the radar. I'll try to have a look if/when the next pipeline fails.

**Comment by adrien-berchet** [comment]
> > Hi @matz-e Do you have any idea why the `proj` package fails? I did not change anything to it.
> 
> Oops, sorry, fell off the radar. I'll try to have a look if/when the next pipeline fails.

No problem :-)
And the new pipeline failed :'(

**Comment by matz-e** [comment]
I disabled the `curl` variant for `proj`, should build now. Do you need `curl` support for that?

**Comment by adrien-berchet** [comment]
Seems it's not so simple 😜
Btw, I don't think we explicitly need curl.

**Comment by matz-e** [comment]
```
--------------------------------------------------------------------------------------------------------------------------
Module /gpfs/bbp.cscs.ch/ssd/apps/bsd/pulls/2133/config/modules/2023-11/applications/linux-rhel7-skylake/py-bluepymm/0.9.5
--------------------------------------------------------------------------------------------------------------------------
bluepymm 0.9.5 has requirement pandas>=2.0.0, but you have pandas 1.5.3.
make: *** [py-bluepymm/0.9.5] Error 1
--------------------------------------------------------------------------------------------------------------------------------
Module /gpfs/bbp.cscs.ch/ssd/apps/bsd/pulls/2133/config/modules/2023-11/applications/linux-rhel7-skylake/py-bluepyparallel/0.0.9
--------------------------------------------------------------------------------------------------------------------------------
bluepyparallel 0.0.9 has requirement ipyparallel<7,>=6.3, but you have ipyparallel 8.4.1.
make: *** [py-bluepyparallel/0.0.9] Error 1
------------------------------------------------------------------------------------------------------------------------------------
Module /gpfs/bbp.cscs.ch/ssd/apps/bsd/pulls/2133/config/modules/2023-11/applications/linux-rhel7-skylake/py-synthesis-workflow/1.0.2
------------------------------------------------------------------------------------------------------------------------------------
atlas-analysis 0.0.6 requires geomdl, which is not installed.
atlas-analysis 0.0.6 requires vtk, which is not installed.
neurocollage 0.3.1 requires trimesh, which is not installed.
synthesis-workflow 1.0.2 requires placement-algorithm, which is not installed.
neurocollage 0.3.1 has requirement bluepysnap<1,>=0.13.1, but you have bluepysnap 1.0.7.
bluepyparallel 0.0.9 has requirement ipyparallel<7,>=6.3, but you have ipyparallel 8.4.1.
bluepymm 0.9.5 has requirement pandas>=2.0.0, but you have pandas 1.5.3.
synthesis-workflow 1.0.2 has requirement matplotlib>=3.6.3, but you have matplotlib 3.6.2.
synthesis-workflow 1.0.2 has requirement neurots<3.4.1,>=3.3.1, but you have neurots 3.4.2.
synthesis-workflow 1.0.2 has requirement region-grower<1.1.1,>=1.0.0, but you have region-grower 1.2.8.
```

What features of the newer `matplotlib` and `pandas` are required?

And why no compatibility with modern `bluepysnap`?

**Comment by adrien-berchet** [comment]
> What features of the newer `matplotlib` and `pandas` are required?

None, I fixed this.

> And why no compatibility with modern `bluepysnap`?

What do you mean?
EDIT: Ah it's in NeuroCollage... I don't know why we don't support newer versions here, I will check that.

**Comment by adrien-berchet** [comment]
Hi @matz-e 
I fixed the dependencies to fix versions of `pandas`, `bluepysnap` and `matplotlib` but now I don't understand the last error about `py-ipyparallel`... Could you help me please? :blush:

**Comment by matz-e** [comment]
> Hi @matz-e I fixed the dependencies to fix versions of `pandas`, `bluepysnap` and `matplotlib` but now I don't understand the last error about `py-ipyparallel`... Could you help me please? 😊

On it. Mainly manually bisecting the dependencies. I disabled the restriction on `py-ipyparallel`, but that just gave more errors about `py-setuptools` :see_no_evil:

File: bluebrain/repo-bluebrain/packages/py-bluepyparallel/package.py [hunk 2]
```diff
@@ -12,15 +12,21 @@ class PyBluepyparallel(PythonPackage):
...
     depends_on("py-mpi4py@3.0.3:", type="run")
     depends_on("py-tqdm@4.28.1:", type="run")
-    depends_on("py-sqlalchemy@:1.3", type="run")
-    depends_on("py-sqlalchemy-utils@0.36:", type="run")
-    depends_on("py-ipyparallel@6.3:", type="run")
+    depends_on("py-sqlalchemy@:1.3", type="run", when="@:0.0.5")
+    depends_on("py-sqlalchemy@:1.4", type="run", when="@0.0.8:")
+    depends_on("py-sqlalchemy-utils@0.36:", type="run", when="@:0.0.5")
+    depends_on("py-sqlalchemy-utils@0.37.2:", type="run", when="@0.0.8:")
+    depends_on("py-ipyparallel@6.3:6", type="run")
```
Comment (matz-e) [inline bluebrain/repo-bluebrain/packages/py-bluepyparallel/package.py]: This is going to be a consistent headache. `py-ipyparallel` < 8 will depend on ancient `py-setuptools`, conflicting with a lot of other packages. Any chance to remove that requirements from `bluepyparallel`?

Comment (adrien-berchet) [reply bluebrain/repo-bluebrain/packages/py-bluepyparallel/package.py]: Ah ok, I see, I'm going to check if I can relax this req. Thanks!

**Comment by adrien-berchet** [comment]
Hi @matz-e , sorry I'm blocked again :disappointed: 
I think I fixed most of the dependencies but I have an issue with vtk now... It seems that CMake can't find `nlohmann_json` but adding it in dependencies does not solve the issue.

**Comment by matz-e** [comment]
> Hi @matz-e , sorry I'm blocked again 😞 I think I fixed most of the dependencies but I have an issue with vtk now... It seems that CMake can't find `nlohmann_json` but adding it in dependencies does not solve the issue.

I'm trying to reproduce that :sweat_smile:, it's compiling away.

File: var/spack/repos/builtin/packages/py-dictdiffer/package.py [hunk 3]
```diff
@@ -12,6 +12,7 @@ class PyDictdiffer(PythonPackage):
     homepage = "https://github.com/inveniosoftware/dictdiffer"
     pypi = "dictdiffer/dictdiffer-0.8.1.tar.gz"
 
+    version("0.9.0", sha256="17bacf5fbfe613ccf1b6d512bd766e6b21fb798822a133aa86098b8ac9997578")
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/py-dictdiffer/package.py:15]: Just for our peace of mind, these modifications in `builtin` packages come from upstream? (@heerener for awareness)

**Review by matz-e — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-dictdiffer/package.py (hunk 3 above)
Comment (heerener) [reply var/spack/repos/builtin/packages/py-dictdiffer/package.py:15]: The modified ones do, but the new versions don't exist in `upstream/develop`.
`py-geomodl` does not exist upstream.

@adrien-berchet  I suggest filing a pull request to https://github.com/spack/spack from a personal fork of that repo with the changes to `var/spack/repos/builtin/packages`. There's some documentation on how to do that here: https://github.com/BlueBrain/spack/blob/develop/bluebrain/documentation/upstreaming.md , but by all means, feel free to reach out if things are unclear / need further documentation!

**Comment by adrien-berchet** [comment]
Ok, I will check that,
Thank you very much for your help on this!

**Comment by adrien-berchet** [comment]
See this PR to add it in the upstream repo: https://github.com/spack/spack/pull/40933

File: var/spack/repos/builtin/packages/py-dictdiffer/package.py (hunk 3 above)
Comment (adrien-berchet) [reply var/spack/repos/builtin/packages/py-dictdiffer/package.py:15]: Should I wait for https://github.com/spack/spack/pull/40933 to be merged or we can already merge this one?

Comment (heerener) [reply var/spack/repos/builtin/packages/py-dictdiffer/package.py:15]: We can probably merge now, then file a new PR to backport the upstream modifications later.

I have some time, so if you want I can file upstream PRs for the version bumps.

Comment (adrien-berchet) [reply var/spack/repos/builtin/packages/py-dictdiffer/package.py:15]: Great, thanks!

Comment (adrien-berchet) [reply var/spack/repos/builtin/packages/py-dictdiffer/package.py:15]: @heerener https://github.com/spack/spack/pull/40933 has been merged in upstream, is there anything else I should do?

Comment (heerener) [reply var/spack/repos/builtin/packages/py-dictdiffer/package.py:15]: As there were no further modifications I think you're good!

Comment (adrien-berchet) [reply var/spack/repos/builtin/packages/py-dictdiffer/package.py:15]: Perfect, thanks!

**Pull request merged**